[docs] Explain multiclock use and restrictions (#528)
firesim/firesim: docs/Advanced-Usage/Generating-Different-Targets.rst

transformed and thus used in FireSim:

#. The top-level module must have no inputs or outputs. Input stimulus and output capture must be 
   implemented using target RTL or target-to-host Bridges.
#. All target clocks must be generated by a single ``RationalClockBridge``.
#. Black boxes must be "clock-gateable" by replacing its input clock with a gated equivalent which will be used
   to stall simulation time in that module.

>    a. As a consequence, target clock-gating cannot be implemented using black-box primitives, and must instead be modelled by

Reviewer: ```suggestion
   a. As a consequence, target clock-gating cannot be implemented using black-box primitives, and must instead be modeled by
```

>       adding clock-enables to all state elements of the gated clock domain (i.e., by adding a feedback mux on registers to

Reviewer: ```suggestion
      adding clock-enables to all state elements of the gated clock domain (i.e., by adding an enable or feedback mux on registers to
```

#. This allows any Verilog module, subject to the constraint above, to be instantiated anywhere in the target
   design using the standard Chisel Blackbox interface.

----------------------------
Multiple Clock Domains
----------------------------

FireSim can support simulating targets that have multiple clock
domains.  As stated above, all clocks must be generated using a single

> ``RationalClockBridge``. For most users the default FireSim test harness in Chipyard will suffice,

Reviewer 2: > For most users the default FireSim test harness in Chipyard will suffice
@davidbiancolin  does this mean that the default `FireSim` test harness can be configured to provide multiple clocks?  Given that you are making this statement in a section called "Multiple Clock Domains", that's what I start thinking but looking at the FireSim 1.10 code, it looks like people who have multiple clock domains still need to look at the `FireSimMulticlockPOC`.

Am I missing something or just confused?

Author: Yeah this is confusing. I think at time of writing it was still rolled into the default harness. We've since elected to split it out into the Proof-of-Concept harness while we figure out how to generalize multiclock targets in Chipyard.

Reviewer 2: Ok.  Thanks for confirming.  The FireSimMulticlockPOC makes sense and I see how I can use parts of it to get what I need.

Author: Was there anything else that seemed unclear?

Reviewer 2: In the rst, no.  Everything else made sense and was clearly written.  

It would be nice if the reference to the source resolved to a link somehow.  I think I've seen projects that use rst make that happen.  If you guys don't have an easy way to do it already, I can go look for the plugin.

if you need to define a custom test harness instantiate the ``RationalClockBridge`` like so:

.. literalinclude:: ../../sim/src/main/scala/midasexamples/TrivialMulticlock.scala
    :language: scala
    :start-after: DOC include start: RationalClockBridge Usage
    :end-before: DOC include end: RationalClockBridge Usage

Further documentation can be found in the source
(``sim/midas/src/main/scala/midas/widgets/ClockBridge.scala``).

=================
The Base Clock
=================
By convention, target time is specified in cycles of the `base clock`,
which is defined to be the zeroth output clock of the ``RationalClockBridge``.
While we suggest making the base clock the fastest clock in your system, as in any
microprocessor-based system it will likely correspond to your core clock
frequency, this is not a constraint.

============
Limitations:
============

> * The number of target clocks FireSim can simulate is bound by the number of BUFGCE resources

Reviewer: ```suggestion
* The number of target clocks FireSim can simulate is bounded by the number of BUFGCE resources
```

>   available on the host-FPGA, as these are used to independently clock-gate each target clock.

Reviewer: ```suggestion
  available on the host FPGA, as these are used to independently clock-gate each target clock.
```

> * As its name suggests, the ``RationalClockBridge`` can only generate target-clocks

Reviewer: ```suggestion
* As its name suggests, the ``RationalClockBridge`` can only generate target clocks
```

>   that are rationally related. Specifically, all requested frequencies must

Reviewer 3: ```suggestion
  that are rationally related. Specifically, all requested frequencies must be
```

  expressable in the form:

  .. math::
    f_{i} = \frac{f_{lcm}}{k_{i}}

  Where,
    * :math:`f_{i}` is the desired frequency of the :math:`i^{th}` clock
    * :math:`f_{lcm}`, is the least-common multiple of all requested frequencies
    * :math:`k_{i}` is a 16-bit unsigned integer

  An arbitrary frequency can be modeled using a sufficiently precise rational
  multiple. Golden Gate will raise a compile-time error if it cannot support
  a desired frequency.

> * Bridge modules must reside in a single clock domain. The Bridge's target-interface

Reviewer: ```suggestion
* Each bridge module must reside entirely within a single clock domain. The Bridge's target interface
```

>   must contain a single input clock, and all inputs and outputs to the

Reviewer: ```suggestion
  must contain a single input clock, and all inputs and outputs of the
```

>   bridge module must be launched and latched, respectively, by registers in the

Reviewer: ```suggestion
  bridge module must be latched and launched, respectively, by registers in the
```
